openmls/src/group/tests/test_proposal_validation.rs

```diff
@@ -1901,3 +1902,149 @@ fn test_valsem112(ciphersuite: Ciphersuite, backend: &impl OpenMlsCryptoProvider
         .expect("Unexpected error.");
 }
+
+// --- PreSharedKey Proposals ---
+
+#[apply(ciphersuites_and_backends)]
+fn test_valsem401_valsem402(ciphersuite: Ciphersuite, backend: &impl OpenMlsCryptoProvider) {
+    let ProposalValidationTestSetup {
+        mut alice_group,
+        alice_credential_with_key_and_signer,
+        mut bob_group,
+        ..
+    } = validation_test_setup(PURE_PLAINTEXT_WIRE_FORMAT_POLICY, ciphersuite, backend);
+
+    let alice_backend = OpenMlsRustCrypto::default();
+    let bob_backend = OpenMlsRustCrypto::default();
+
+    let bad_psks = [
+        // ValSem401
+        (
+            vec![PreSharedKeyId::external(
+                b"irrelevant".to_vec(),
+                zero(ciphersuite.hash_length() + 1),
+            )],
+            ProcessMessageError::InvalidCommit(StageCommitError::ProposalValidationError(
+                ProposalValidationError::Psk(PskError::NonceLengthMismatch {
+                    expected: ciphersuite.hash_length(),
+                    got: ciphersuite.hash_length() + 1,
+                }),
+            )),
+        ),
+        // ValSem401
+        (
+            vec![PreSharedKeyId::external(
+                b"irrelevant".to_vec(),
+                zero(ciphersuite.hash_length() - 1),
+            )],
+            ProcessMessageError::InvalidCommit(StageCommitError::ProposalValidationError(
+                ProposalValidationError::Psk(PskError::NonceLengthMismatch {
+                    expected: ciphersuite.hash_length(),
+                    got: ciphersuite.hash_length() - 1,
+                }),
+            )),
+        ),
+        // ValSem402
+        (
+            vec![PreSharedKeyId::resumption(
+                ResumptionPskUsage::Reinit,
+                alice_group.group_id().clone(),
+                alice_group.epoch(),
+                zero(ciphersuite.hash_length()),
+            )],
+            ProcessMessageError::InvalidCommit(StageCommitError::ProposalValidationError(
+                ProposalValidationError::Psk(PskError::UsageMismatch {
+                    allowed: vec![ResumptionPskUsage::Application],
+                    got: ResumptionPskUsage::Reinit,
+                }),
+            )),
+        ),
+        // ValSem402
+        (
+            vec![PreSharedKeyId::resumption(
+                ResumptionPskUsage::Branch,
+                alice_group.group_id().clone(),
+                alice_group.epoch(),
+                zero(ciphersuite.hash_length()),
+            )],
+            ProcessMessageError::InvalidCommit(StageCommitError::ProposalValidationError(
+                ProposalValidationError::Psk(PskError::UsageMismatch {
+                    allowed: vec![ResumptionPskUsage::Application],
+                    got: ResumptionPskUsage::Branch,
+                }),
+            )),
+        ),
+        // Note: This can't be tested because duplicate proposals are filtered.
```

Comment:
These commented out tests should also get the issue referenced that will look into whether we need to do this or not.